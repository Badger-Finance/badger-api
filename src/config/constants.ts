import { ethers } from 'ethers';

const RAW_TOKENS = {
  // eth tokens
  BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
  DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',
  WBTC: '0x2260fac5e5542a773aa44fbcfedf7c193bc2c599',
  WETH: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
  SUSHI: '0x6b3595068778dd592e39a122f4f5a5cf09c90fe2',

  // bsc tokens
  CAKE: '0x0E09FaBB73Bd3Ade0a17ECC321fD13a19e81cE82',
  WBNB: '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c',
  BTCB: '0x7130d2A12B9BCbFAe4f2634d864A1Ee1Ce3Ead9c',
  BSC_BADGER: '0x753fbc5800a8C8e3Fb6DC6415810d627A387Dfc9',
  BSC_BBADGER: '0x1F7216fdB338247512Ec99715587bb97BBf96eae',
  BSC_BDIGG: '0x5986D5c77c65e5801a5cAa4fAE80089f870A71dA',

  // curve tokens
  CRV_RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
  CRV_TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
  CRV_SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',

  // uniswap tokens
  UNI_BADGER_WBTC: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
  UNI_DIGG_WBTC: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

  // sushiswap tokens
  SUSHI_ETH_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
  SUSHI_BADGER_WBTC: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
  SUSHI_DIGG_WBTC: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',

  // pancakeswap tokens
  PANCAKE_BNB_BTCB: '0x356b7d0d3c54F22C82B7a670C6Ba9E2381b0624c',
  PANCAKE_BBADGER_BTCB: '0x87Ae7b5c43D4e160cDB9427a78BA87B9503ee37b',
  PANCAKE_BDIGG_BTCB: '0xfbfa92e037e37F946c0105902640914E3aCe6752',

  // eth vault tokens
  BBADGER: '0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28',
  BDIGG: '0x7e7e112a68d8d2e221e11047a72ffc1065c38e1a',
  BUNI_BADGER_WBTC: '0x235c9e24d3fb2fafd58a2e49d454fdcd2dbf7ff1',
  BUNI_DIGG_WBTC: '0xc17078fdd324cc473f8175dc5290fae5f2e84714',
  BSUSHI_ETH_WBTC: '0x758a43ee2bff8230eeb784879cdcff4828f2544d',
  BSUSHI_BADGER_WBTC: '0x1862a18181346ebd9edaf800804f89190def24a5',
  BSUSHI_DIGG_WBTC: '0x88128580acdd9c04ce47afce196875747bf2a9f6',
  BCRV_SBTC: '0xd04c48a53c111300ad41190d63681ed3dad998ec',
  BCRV_RENBTC: '0x6def55d2e18486b9ddfaa075bc4e4ee0b28c1545',
  BCRV_TBTC: '0xb9d076fde463dbc9f915e5392f807315bf940334',
  BCRV_HRENBTC: '0xaf5a1decfa95baf63e0084a35c62592b774a2a87',
  BVYWBTC: '0x4b92d19c11435614cd49af1b589001b7c08cd4d5',

  // bsc vault tokens
  BPANCAKE_BNB_BTCB: '0xaf4B9C4b545D5324904bAa15e29796D2E2f90813',
  BPANCAKE_BBADGER_BTCB: '0x857F91f735f4B03b19D2b5c6E476C73DB8241F55',
  BPANCAKE_BDIGG_BTCB: '0xa861Ba302674b08f7F2F24381b705870521DDfed',
};

export const TOKENS = Object.fromEntries(
  Object.entries(RAW_TOKENS).map(([key, val]) => [key, ethers.utils.getAddress(val)]),
);

export enum Protocol {
  Curve = 'curve',
  Sushiswap = 'sushiswap',
  Uniswap = 'uniswap',
  Pancakeswap = 'pancakeswap',
  Yearn = 'yearn',
}

// data point constants - index two times per hour, 48 per day
export const CURRENT = 0;
export const ONE_DAY = 24 * 2;
export const ONE_MINUTE_MS = 60 * 1000;
export const THREE_DAYS = ONE_DAY * 3;
export const SEVEN_DAYS = ONE_DAY * 7;
export const THIRTY_DAYS = ONE_DAY * 30;
export const SAMPLE_DAYS = THIRTY_DAYS + 1;
export const ONE_YEAR_MS = 365 * 24 * 60 * 60 * 1000;

// data access constants
export const APY_SNAPSHOTS_DATA = process.env.APY_SNAPSHOTS_DATA || 'MISSING REQUIRED ENV VAR';
export const ASSET_DATA = process.env.ASSET_DATA || 'MISSING REQUIRED ENV VAR';
export const PRICE_DATA = process.env.PRICE_DATA || 'MISSING REQUIRED ENV VAR';
export const SETT_SNAPSHOTS_DATA = process.env.SETT_SNAPSHOTS_DATA || 'MISSING REQUIRED ENV VAR';
export const REWARD_DATA = process.env.REWARD_DATA || 'MISSING REQUIRED ENV VAR';
export const UNISWAP_URL = process.env.UNISWAP || 'MISSING REQUIRED ENV VAR';
export const SUSHISWAP_URL = process.env.SUSHISWAP || 'MISSING REQUIRED ENV VAR';
export const PANCAKESWAP_URL = process.env.PANCAKESWAP || 'MISSING REQUIRED ENV VAR';
export const BADGER_URL = process.env.BADGER || 'MISSING REQUIRED ENV VAR';
export const BADGER_BSC_URL = process.env.BADGER_BSC || 'MISSING REQUIRED ENV VAR';
export const BADGER_DAO_URL = process.env.BADGER_DAO || 'MISSING REQUIRED ENV VAR';
export const MASTERCHEF_URL = process.env.MASTERCHEF || 'MISSING REQUIRED ENV VAR';
export const BOUNCER_PROOFS = process.env.BOUNCER_PROOFS || 'MISSING REQUIRED ENV VAR';

// third party api constants
export const CURVE_API_URL = 'https://stats.curve.fi/raw-stats/apys.json';
export const COINGECKO_URL = 'https://api.coingecko.com/api/v3/simple';

export enum Provider {
  Cloudflare = 'https://cloudflare-eth.com/',
  Binance = 'https://bsc-dataseed.binance.org/',
  MyEtherWallet = 'https://nodes.mewapi.io/rpc/eth',
}
export const SUSHI_CHEF = '0xc2EdaD668740f1aA35E4D8f227fB8E17dcA888Cd';
export const PANCAKE_CHEF = '0x73feaa1eE314F8c655E354234017bE2193C9E24E';
