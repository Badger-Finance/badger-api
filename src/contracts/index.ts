/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
export type { BalancerVault } from "./BalancerVault";
export type { CurvePool } from "./CurvePool";
export type { CurvePool3 } from "./CurvePool3";
export type { CvxLocker } from "./CvxLocker";
export type { SolidlyPair } from "./SolidlyPair";
export type { StablePhantomVault } from "./StablePhantomVault";
export type { StablePool } from "./StablePool";
export type { UniV2 } from "./UniV2";
export type { WeightedPool } from "./WeightedPool";
export * as factories from "./factories";
export { BalancerVault__factory } from "./factories/BalancerVault__factory";
export { CurvePool3__factory } from "./factories/CurvePool3__factory";
export { CurvePool__factory } from "./factories/CurvePool__factory";
export { CvxLocker__factory } from "./factories/CvxLocker__factory";
export { SolidlyPair__factory } from "./factories/SolidlyPair__factory";
export { StablePhantomVault__factory } from "./factories/StablePhantomVault__factory";
export { StablePool__factory } from "./factories/StablePool__factory";
export { UniV2__factory } from "./factories/UniV2__factory";
export { WeightedPool__factory } from "./factories/WeightedPool__factory";
