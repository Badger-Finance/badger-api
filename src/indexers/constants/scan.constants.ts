export const MAX_SCAN_RANGE = 300_000;
