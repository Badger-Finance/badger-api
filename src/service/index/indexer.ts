import {
	EventInput,
	getBlock,
	getIndexedBlock,
	GetPriceFunc,
	getSett,
	respond,
	saveItem,
	THIRTY_MIN_BLOCKS,
} from '../util/util';

const ASSET_DATA = process.env.ASSET_DATA || ''; // FIXME: sane defaults?

export const indexAsset = async (event: EventInput, getPrice: GetPriceFunc) => {
	const { asset, createdBlock, contract } = event;
	let block = await getIndexedBlock(ASSET_DATA, asset, createdBlock);
	console.log(`Index ${asset} at height: ${block}`);

	while (true) {
		const sett = await getSett(contract, block);

		if (sett.errors != undefined) {
			break;
		}

		if (sett.data == null || sett.data.sett == null) {
			block += THIRTY_MIN_BLOCKS;
			continue;
		}

		const { balance, totalSupply, pricePerFullShare } = sett.data.sett;

		const decimals = asset === 'digg' ? 9 : 18;
		const blockData = await getBlock(block);
		const timestamp = Number(blockData.timestamp) * 1000;
		const balanceComputed = balance / Math.pow(10, decimals);
		const supply = totalSupply / Math.pow(10, decimals);
		const ratio = pricePerFullShare / Math.pow(10, 18);
		const value = balanceComputed * (await getPrice(sett));

		const snapshot = {
			asset: asset,
			height: block,
			timestamp: timestamp,
			balance: balance,
			supply: supply,
			ratio: ratio,
			value: parseFloat(value.toFixed(2)),
		};

		await saveItem(ASSET_DATA, snapshot);
		block += THIRTY_MIN_BLOCKS;
	}

	return respond(200); // FIXME: is this right?
};
