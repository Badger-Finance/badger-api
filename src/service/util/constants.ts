export const TOKENS = {
	// general tokens
	RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
	BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
	TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
	SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',
	DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',

	// uniswap tokens
	UNI_BADGER: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
	UNI_DIGG: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

	// sushiswap tokens
	SUSHI_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
	SUSHI_BADGER: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
	SUSHI_DIGG: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',
};
