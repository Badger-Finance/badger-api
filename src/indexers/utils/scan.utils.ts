import { Network } from '@badger-dao/sdk';
import { BlocksRangeOptions } from '@badger-dao/sdk/lib/common';

import { getDataMapper } from '../../aws/dynamodb.utils';
import { IndexingMetadata } from '../../aws/models/indexing-metadata.model';
import { MAX_SCAN_RANGE } from '../constants/scan.constants';
import { LastScannedBlockMeta } from '../interfaces/last-scanned-block-meta.interface';

export async function getOrCreateMetadata<T>(task: string): Promise<IndexingMetadata<T>> {
  const mapper = getDataMapper();

  let metadata!: IndexingMetadata<T>;

  try {
    metadata = await mapper.get(Object.assign(new IndexingMetadata<T>(), { task }));
  } catch (e) {
    metadata = Object.assign(new IndexingMetadata<T>(), { task, data: {} });
  }

  return metadata;
}

export function getLastScanedBlock<T extends LastScannedBlockMeta>(
  indexingMetaData: T,
  network: Network,
  address: string,
) {
  return indexingMetaData[network]?.[address]?.lastScannedBlock || 0;
}

export function getScanRangeOpts(
  fromChain: Required<BlocksRangeOptions>,
  lastScannedBlock: number,
): Required<BlocksRangeOptions> {
  const startBlock = lastScannedBlock > 0 ? lastScannedBlock + 1 : fromChain.startBlock;
  let endBlock = fromChain.endBlock;

  if (endBlock - startBlock > MAX_SCAN_RANGE) {
    endBlock = startBlock + MAX_SCAN_RANGE;
  }

  return {
    startBlock,
    endBlock,
  };
}

export async function saveIndexingMetadata<T>(model: IndexingMetadata<T>): Promise<void> {
  const mapper = getDataMapper();

  try {
    await mapper.put(model);
  } catch (e) {
    console.error(`Failed to save indexing metadata for ${model.task}`, e);
  }
}
