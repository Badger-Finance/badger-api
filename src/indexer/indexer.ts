import { ASSET_DATA } from '../util/constants';
import {
	EventInput,
	getBlock,
	getIndexedBlock,
	GetPriceFunc,
	getSett,
	saveItem,
	THIRTY_MIN_BLOCKS,
} from '../util/util';

export const indexAsset = async (event: EventInput, getPrice: GetPriceFunc) => {
	const { asset, createdBlock, contract } = event;
	let block = await getIndexedBlock(ASSET_DATA, asset, createdBlock);
	console.log(`Index ${asset} at height: ${block}`);

	while (true) {
		const sett = await getSett(contract, block);

		if (sett.errors != undefined) {
			break;
		}

		if (sett.data == null || sett.data.sett == null) {
			block += THIRTY_MIN_BLOCKS;
			continue;
		}

		const { balance, totalSupply, pricePerFullShare, token } = sett.data.sett;

		const blockData = await getBlock(block);
		const timestamp = Number(blockData.timestamp) * 1000;
		const tokenBalance = balance / Math.pow(10, token.decimals);
		const supply = totalSupply / Math.pow(10, token.decimals);
		const ratio = pricePerFullShare / Math.pow(10, 18);
		const value = tokenBalance * (await getPrice(sett)).usd;

		const snapshot = {
			asset: asset,
			height: block,
			timestamp: timestamp,
			balance: balance,
			supply: supply,
			ratio: ratio,
			value: parseFloat(value.toFixed(2)),
		};

		await saveItem(ASSET_DATA, snapshot);
		block += THIRTY_MIN_BLOCKS;
	}
};
