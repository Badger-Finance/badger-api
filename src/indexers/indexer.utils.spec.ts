import {
  BadgerGraph,
  EmissionControl,
  EmissionControl__factory,
  RegistryVault,
  Strategy__factory,
  VaultsService,
  VaultV15,
  VaultV15__factory,
} from '@badger-dao/sdk';
import { Strategy } from '@badger-dao/sdk/lib/contracts/Strategy';
import * as gqlGenT from '@badger-dao/sdk/lib/graphql/generated/badger';
import graphVaults from '@badger-dao/sdk-mocks/generated/ethereum/graph/loadSetts.json';
import vaultStrategy from '@badger-dao/sdk-mocks/generated/ethereum/vaults/getVaultStrategy.json';
import registryVaults from '@badger-dao/sdk-mocks/generated/ethereum/vaults/loadVaults.json';
import { BigNumber } from 'ethers';

import { TestEthereum } from '../chains/config/test.config';
import { MOCK_VAULT, MOCK_VAULT_DEFINITION, TEST_CURRENT_TIMESTAMP } from '../test/constants';
import { setupMockChain } from '../test/mocks.utils';
import { mockContract } from '../test/mocks.utils/contracts/mock.contract.base';
import * as apiVaults from '../vaults/vaults.service';
import { constructVaultDefinition, vaultToSnapshot } from './indexer.utils';

describe('indexer.utils', () => {
  async function setupBaseUtilFixture() {
    jest.spyOn(Date, 'now').mockImplementation(() => TEST_CURRENT_TIMESTAMP);

    jest.spyOn(console, 'error').mockImplementation();
    jest.spyOn(console, 'warn').mockImplementation();

    const strategyMock = mockContract<Strategy>(Strategy__factory);
    const vaultV15Mock = mockContract<VaultV15>(VaultV15__factory);
    const emissionControl = mockContract<EmissionControl>(EmissionControl__factory);

    jest.spyOn(strategyMock, 'withdrawalFee').mockImplementation(async () => BigNumber.from(1));
    jest.spyOn(strategyMock, 'performanceFeeGovernance').mockImplementation(async () => BigNumber.from(2));
    jest.spyOn(strategyMock, 'performanceFeeStrategist').mockImplementation(async () => BigNumber.from(3));

    jest.spyOn(vaultV15Mock, 'withdrawalFee').mockImplementation(async () => BigNumber.from(1));
    jest.spyOn(vaultV15Mock, 'performanceFeeGovernance').mockImplementation(async () => BigNumber.from(2));
    jest.spyOn(vaultV15Mock, 'performanceFeeStrategist').mockImplementation(async () => BigNumber.from(3));
    jest.spyOn(vaultV15Mock, 'managementFee').mockImplementation(async () => BigNumber.from(4));

    jest.spyOn(emissionControl, 'boostedEmissionRate').mockImplementation(async () => BigNumber.from(1));

    const chain = setupMockChain();

    jest.spyOn(VaultsService.prototype, 'loadVault').mockImplementation(async function ({ address }) {
      return <RegistryVault>registryVaults.find((v) => v.address === address);
    });
    jest.spyOn(VaultsService.prototype, 'getVaultStrategy').mockImplementation(async () => vaultStrategy);

    jest.spyOn(BadgerGraph.prototype, 'loadSett').mockImplementation(async ({ id }) => {
      const graphVault = graphVaults.setts.find((v) => v.id === id);

      if (!graphVault) return { sett: null };

      return { sett: <gqlGenT.SettQuery['sett']>graphVault, __typename: 'Query' };
    });

    jest.spyOn(apiVaults.VaultsService, 'loadVault').mockImplementation(async () => MOCK_VAULT);

    return { chain, strategyMock, vaultV15Mock };
  }

  describe('vaultToSnapshot', () => {
    let testChain: TestEthereum;

    beforeEach(async () => {
      const { chain } = await setupBaseUtilFixture();

      testChain = chain;
    });

    it('should return valid snapshot of the vault', async () => {
      const snapshot = await vaultToSnapshot(testChain, MOCK_VAULT_DEFINITION);

      // small lol here
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('constructVaultsDefinition', () => {
    let testChain: TestEthereum;

    beforeEach(async () => {
      const { chain } = await setupBaseUtilFixture();

      testChain = chain;
    });

    it('should construct vault definition', async () => {
      expect(await constructVaultDefinition(testChain, <RegistryVault>registryVaults[0])).toMatchSnapshot();
    });

    it('should return null if no data on the Graph', async () => {
      jest.spyOn(BadgerGraph.prototype, 'loadSett').mockImplementation(async () => ({ sett: null }));

      expect(await constructVaultDefinition(testChain, <RegistryVault>registryVaults[1])).toBeNull();
    });
  });
});
