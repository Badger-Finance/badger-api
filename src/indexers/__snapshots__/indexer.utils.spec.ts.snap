// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indexer.utils constructVaultsDefinition should construct vault definition 1`] = `
VaultDefinitionModel {
  "address": "0xd04c48A53c111300aD41190D63681ed3dAd998eC",
  "behavior": "None",
  "bouncer": "none",
  "chain": "ethereum",
  "client": "",
  "createdAt": NaN,
  "depositToken": "0x075b1bb99792c9E1041bA13afEf80C91a1e70fB3",
  "id": "ethereum-0xd04c48A53c111300aD41190D63681ed3dAd998eC",
  "isNew": false,
  "isProduction": 1,
  "name": "renBTC/wBTC/sBTC",
  "protocol": "Convex",
  "releasedAt": NaN,
  "stage": "prod",
  "state": "discontinued",
  "updatedAt": NaN,
  "version": "v1",
}
`;

exports[`indexer.utils getLpTokenBalances should return token balances of the vault 1`] = `
VaultTokenBalance {
  "tokenBalances": Array [
    Object {
      "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
      "balance": 0,
      "decimals": 18,
      "name": "Badger",
      "symbol": "BADGER",
      "value": 0,
    },
    Object {
      "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
      "balance": 0,
      "decimals": 8,
      "name": "Wrapped BTC",
      "symbol": "WBTC",
      "value": 0,
    },
  ],
  "vault": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
}
`;

exports[`indexer.utils getVault should return graph vault sequence 1`] = `
Object {
  "__typename": "Query",
  "sett": Object {
    "available": "1213945495264444893468113",
    "balance": "1213945495264444893468113",
    "controller": Object {
      "id": "0x63cf44b2548e4493fd099222a1ec79f3344d9682",
    },
    "decimals": "18",
    "grossDeposit": "18406633496620429219098123",
    "grossShareDeposit": "15571214606848692106223327",
    "grossShareWithdraw": "14588168614349076090183817",
    "grossWithdraw": "17527900740448259273124826",
    "id": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
    "name": "Badger Sett Badger",
    "netDeposit": "878732756172169945973297",
    "netShareDeposit": "983045992499616016039510",
    "pricePerFullShare": "1234881688676350581",
    "strategy": Object {
      "balance": "0",
      "id": "0x75b8e21bd623012efb3b69e1b562465a68944ee6",
    },
    "symbol": "bBADGER",
    "token": Object {
      "decimals": "18",
      "id": "0x3472a5a71965499acd81997a54bba8d852c6e53d",
      "name": "Badger",
      "symbol": "BADGER",
      "totalSupply": "21000000000000000000000000",
    },
    "totalSupply": "983045992499616016039510",
  },
}
`;

exports[`indexer.utils vaultToSnapshot should return valid snapshot of the vault 1`] = `
Object {
  "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "apr": 0.3067433490728799,
  "available": 1006655.3375347335,
  "balance": 1006655.3375347335,
  "block": 13500500,
  "boostWeight": 0,
  "harvestApr": 0.3067433490728799,
  "pricePerFullShare": 1.2348816886763505,
  "strategy": Object {
    "address": "0x32BdF2B35Bbf8dAca3561C3ADE59c941488E9c69",
    "aumFee": 0,
    "performanceFee": 2,
    "strategistFee": 3,
    "withdrawFee": 1,
  },
  "strategyBalance": 0,
  "timestamp": 1660223160457,
  "totalSupply": 815183.6299506156,
  "value": 13515354561.74,
  "yieldApr": 0.3067433490728799,
}
`;
